qiskit_aer/backends/aer_compiler.py

```diff
@@ -199,12 +215,78 @@ def _inline_circuit(self, circ, continue_label, break_label, bit_map=None):
+    class _ClbitConverter(ExprVisitor):
+        """Apply cbit index conversion in Expr tree"""
+
+        def __init__(self, bit_map):
+            self.bit_map = bit_map
+
+        def visit_value(self, node, /):
+            # pylint: disable=unused-variable
+            return node
+
+        def visit_var(self, node, /):
+            if isinstance(node.var, Clbit):
+                node.var = self.bit_map[node.var]
+            else:
+                node.var = [self.bit_map[clbit] for clbit in node.var]
+            return node
+
+        def visit_cast(self, node, /):
+            node.operand.accept(self)
+            return node
+
+        def visit_unary(self, node, /):
+            node.operand.accept(self)
+            return node
+
+        def visit_binary(self, node, /):
+            node.left.accept(self)
+            node.right.accept(self)
+            return node
+
+        def visit_generic(self, node, /):
+            raise AerError(f"unsupported expression is used: {node.__class__}")
+
+    def _convert_jump_conditional(self, cond_tuple, bit_map):
         """Convert a condition tuple according to the wire map."""
-        if isinstance(cond_tuple[0], Clbit):
+        if isinstance(cond_tuple, Expr):
+            return cond_tuple
+        elif isinstance(cond_tuple[0], Clbit):
             return (bit_map[cond_tuple[0]], cond_tuple[1])
-        # ClassicalRegister conditions should already be in the outer circuit.
-        return cond_tuple
+        elif isinstance(cond_tuple[0], ClassicalRegister):
+            # ClassicalRegister conditions should already be in the outer circuit.
+            return cond_tuple
+
+        # left and right expression
+        def _convert_clbit_index(expr):
+            if isinstance(expr, Unary):
+                _convert_clbit_index(expr.operand)
+            elif isinstance(expr, Binary):
+                _convert_clbit_index(expr.left)
+                _convert_clbit_index(expr.right)
+            elif isinstance(expr, Var):
+                if isinstance(expr.var, Clbit):
+                    expr.var = bit_map[expr.var]
+                else:
+                    expr.var = [bit_map[clbit] for clbit in expr.var]
+            return expr
+
+        left = deepcopy(cond_tuple[0]).accept(AerCompiler._ClbitConverter(bit_map))
+        right = cond_tuple[1]
+        return (left, right)
+
+    def _list_clbit_from_expr(self, bit_map, clbits, expr):
+        if isinstance(expr, Unary):
+            self._list_clbit_from_expr(bit_map, clbits, expr.operand)
+        elif isinstance(expr, Binary):
+            self._list_clbit_from_expr(bit_map, clbits, expr.left)
+            self._list_clbit_from_expr(bit_map, clbits, expr.right)
+        elif isinstance(expr, Var):
+            if isinstance(expr.var, Clbit):
+                clbits += bit_map[expr.var]
+            else:
+                clbits += [bit_map[clbit] for clbit in expr.var]
```

Comment:
There's a helper in Terra for iterating only over the `Var` instances: `qiskit.circuit.classical.expr.iter_vars`.  You could implement this like
```python
from qiskit.circuit.classical import expr

def _clbits_from_expr(self, bit_map, node):
    out = set()
    for var in expr.iter_vars(node):
        if isinstance(var.var, Clbit):
            out.add(bit_map[var.var])
        elif isinstance(var.var, ClassicalRegister):
            out.update(bit_map[bit] for bit in var.var)
        else:
            # not handled
    return out
```
and you don't need to manually handle all the possible entries in the tree (for example, your current implementation forgets `Cast`), and Terra will automatically handle any new `Expr` subclasses that get added for you.

Reply:
Thanks. I changed the code as you suggested.